app/controllers/application_controller.rb

```diff
@@ -302,24 +302,16 @@ def generic_exception(exception)
   def set_taxonomy
     return if User.current.nil?
 
-    if SETTINGS[:organizations_enabled]
-      orgs = Organization.my_organizations
-      Organization.current = if orgs.count == 1 && !User.current.admin?
-                               orgs.first
-                             elsif session[:organization_id]
-                               orgs.find_by_id(session[:organization_id])
-                             end
-      warning _("Organization you had selected as your context has been deleted.") if (session[:organization_id] && Organization.current == nil)
-    end
-
-    if SETTINGS[:locations_enabled]
-      locations = Location.my_locations
-      Location.current = if locations.count == 1 && !User.current.admin?
-                           locations.first
-                         elsif session[:location_id]
-                           locations.find_by_id(session[:location_id])
-                         end
-      warning _("Location you had selected as your context has been deleted.") if (session[:location_id] && Location.current == nil)
+    Taxonomy.enabled_taxonomies.each do |taxonomy|
+      next unless SETTINGS[:"#{taxonomy}_enabled"]
+      taxonomy_class = taxonomy.classify.constantize
+      taxonomies = taxonomy_class.send(:"my_#{taxonomy}")
+      taxonomy_class.current = if taxonomies.count == 1 && !User.current.admin?
+                                 taxonomies.first
+                               elsif session[:"#{taxonomy.singularize}_id"]
+                                 taxonomies.find_by_id(session[:"#{taxonomy.singularize}_id"])
+                               end
+      warning _("The %s you had selected as your context has been deleted.") % taxonomy.singularize if (session[:"#{taxonomy.singularize}_id"] && taxonomy_class.current.nil?)
```

Comment:
a) "taxonomy.singularize" isn't translated, it will only ever return an English word
b) could we keep the full strings for both locations and organisations?  It will result in better translations if explicit.  (Though I admit other parts of multi-org have equally poor strings.)
c) tests are failing due to the string change